blog/content/post/2020-12-14-trenchboot_attestation.md

```diff
@@ -0,0 +1,266 @@
+	}
+
+	r = Esys_LoadExternal(ctx, ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE,
+			NULL, external_public_key, TPM2_RH_OWNER, key_handle);
+	if (r != TSS2_RC_SUCCESS ) {
+		charra_log_error("Loading external public key failed.");
+		return CHARRA_RC_ERROR;
+	}
+
+	return CHARRA_RC_SUCCESS;
+}
+```
+
+The following function allows to load external key and it creates `key_handle` that
+is used for the TPM signature verification.
+
+The full scope of changes is available in the following
+[pull request](https://github.com/3mdeb/charra/pull/1).
+
+### Using CHARRA with TPM device
+
+The PoC CHARRA implementation is using the TPM simulator provided in the docker
+container. To use CHARRA with the physical TPM we have created the Yocto layer
+- [meta-trenchboot-attestation](https://github.com/3mdeb/meta-trenchboot-attestation/tree/add_charra),
+which provides the required libraries, that are used by attester and verifier.
+The Yocto Project (YP) is an open source collaboration project that helps
+developers create custom Linux-based systems regardless of the hardware
+architecture. [meta-trenchboot](https://github.com/3mdeb/meta-trenchboot)
+builds the image which contains the TrenchBoot utilities.
+
+### Proof of concept
+
+To proof the following concept we used the ASRock 4x4 Box R1000V with physical
+TPM as the attester and the PC with TPM simulator as the verifier.
+The attester is waiting for the attestation request from the verifier.
+The following logs and videos show the attestation process.
+
+**Attester**
+
+```
+01:05:37 INFO  src/attester.c:73: [attester] Starting up.
+01:05:37 INFO  src/attester.c:82: [attester] Initializing CoAP endpoint.
+01:05:37 INFO  src/attester.c:94: [attester] Registering CoAP resources.
+01:05:37 INFO  src/util/coap_util.c:51: Adding CoAP FETCH resource 'attest'.
+01:05:37 INFO  src/attester.c:101: [attester] Waiting for connections.
+01:06:06 INFO  src/attester.c:129: [attester] Resource 'attest': Received message.
+01:06:06 INFO  src/attester.c:140: [attester] Received data of length 63.
+01:06:06 INFO  src/attester.c:145: [attester] Parsing received CBOR data.
+01:06:06 INFO  src/attester.c:155: [attester] Preparing TPM quote data.
+01:06:06 INFO  src/attester.c:184: [attester] Loading TPM key.
+01:06:06 INFO  src/core/charra_key_mgr.c:36: Loading key "PK.RSA.default".
+01:06:19 INFO  src/util/tpm2_util.c:128: Primary Key created successfully.
+01:06:19 INFO  src/attester.c:193: [attester] Do TPM Quote.
+01:06:19 INFO  src/attester.c:196: [attester] sig_key_handle: 0x40418487
+01:06:19 INFO  src/attester.c:197: [attester] public_key: 1
+01:06:19 INFO  src/attester.c:203: [attester] TPM Quote successful.
+01:06:19 INFO  src/attester.c:209: [attester] Preparing response.
+01:06:19 INFO  src/attester.c:218: [attester] Marshaling response.
+01:06:19 INFO  src/attester.c:224: [attester] Adding marshaled data to CoAP response PDU and send it.
+01:06:19 INFO  src/attester.c:101: [attester] Waiting for connections.
```

Comment:
Some comment below what we can read from these logs would be helpful.

Reply:
Added.